### PR TITLE
autograb (without inputlock, for now)

### PR DESCRIPTION
**Pull request by sunweaver**
@uli42: I took your pr/autograb branch and stripped out the autograb feature only for this PR for now.

The autograb feature is an interesting feature for our current sponsor, so we should have that in nxagent ASAP.

I will add some code, that triggers an NX dialog call when the autograb mode gets enabled. Once that commit is there, I'd love to have you review this subset of commits derived from your pr/autograb branch.

### CODE REVIEW

**Comment by uli42** [comment]
Ok, but the fullscreen problem is still unfixed

Vom Smartphone gesendet.

----- Ursprüngliche Nachricht -----
Von: "Mike Gabriel" <notifications@github.com>
Gesendet: ‎24.‎07.‎2017 14:32
An: "ArcticaProject/nx-libs" <nx-libs@noreply.github.com>
Cc: "Ulrich Sibiller" <uli42@gmx.de>; "Mention" <mention@noreply.github.com>
Betreff: [ArcticaProject/nx-libs] autograb (without inputlock, for now) (#494)

@uli42: I took your pr/autograb branch and stripped out the autograb feature only for this PR for now.
The autograb feature is an interesting feature for our current sponsor, so we should have that in nxagent ASAP.
I will add some code, that triggers an NX dialog call when the autograb mode gets enabled. Once that commit is there, I'd love to have you review this subset of commits derived from your pr/autograb branch.



You can view, comment on, or merge this pull request online at:
  https://github.com/ArcticaProject/nx-libs/pull/494
Commit Summary
nxagent: Add autograb mode. 
nxagent: Refactor AutoGrab. 
nxagent: Make autograb and inputlock nxagent options. 
Args.c: Add command line option -autograb. 
Args.c: make autograb available via options file, too. 
Events.c: improve debugging output for FocusIn/Out 
Events.c: do not ungrab keyboard on LeaveNotify when in autograb mode 
Events.c: Autograb only for windowed modes 
nxagent.1: fix indentation 
File Changes
M doc/nxagent/README.keystrokes (3) 
M etc/keystrokes.cfg (1) 
M nx-X11/programs/Xserver/hw/nxagent/Args.c (22) 
M nx-X11/programs/Xserver/hw/nxagent/Events.c (153) 
M nx-X11/programs/Xserver/hw/nxagent/Events.h (3) 
M nx-X11/programs/Xserver/hw/nxagent/Keystroke.c (9) 
M nx-X11/programs/Xserver/hw/nxagent/Keystroke.h (2) 
M nx-X11/programs/Xserver/hw/nxagent/Options.c (2) 
M nx-X11/programs/Xserver/hw/nxagent/Options.h (6) 
M nx-X11/programs/Xserver/hw/nxagent/man/nxagent.1 (7) 
Patch Links:
https://github.com/ArcticaProject/nx-libs/pull/494.patch 
https://github.com/ArcticaProject/nx-libs/pull/494.diff

**Comment by sunweaver** [comment]
On  Mo 24 Jul 2017 14:38:42 CEST, Ulrich Sibiller wrote:

> Ok, but the fullscreen problem is still unfixed
>

uli42: What fullscreen problem exactly? Can observe any issues here.

Mike
-- 

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

**Comment by uli42** [comment]
Activate autograb, Switch to Fullscreen using keystroke, Switch back . Try the same with switch-all-screens.

For one of those autograb is broken afterwards.

Vom Smartphone gesendet.

----- Ursprüngliche Nachricht -----
Von: "Mike Gabriel" <notifications@github.com>
Gesendet: ‎24.‎07.‎2017 14:58
An: "ArcticaProject/nx-libs" <nx-libs@noreply.github.com>
Cc: "Ulrich Sibiller" <uli42@gmx.de>; "Mention" <mention@noreply.github.com>
Betreff: Re: [ArcticaProject/nx-libs] autograb (without inputlock, for now)(#494)

On Mo 24 Jul 2017 14:38:42 CEST, Ulrich Sibiller wrote:

> Ok, but the fullscreen problem is still unfixed
>

uli42: What fullscreen problem exactly? Can observe any issues here.

Mike
-- 

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22 0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

**Comment by sunweaver** [comment]
On  Mo 24 Jul 2017 15:12:06 CEST, Ulrich Sibiller wrote:

> Activate autograb, Switch to Fullscreen using keystroke, Switch back  
> . Try the same with switch-all-screens.
>
> For one of those autograb ist broken after wards.

@uli42: No, it is not. Tried that using nested i3 desktops. Works as  
expected here.

Just tried it once more. Can't trigger any sort of failure.

Mike
-- 

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

**Comment by uli42** [comment]
So its kwin again... That is even worse. It should work with all WMs equally.

Vom Smartphone gesendet.

----- Ursprüngliche Nachricht -----
Von: "Mike Gabriel" <notifications@github.com>
Gesendet: ‎24.‎07.‎2017 15:16
An: "ArcticaProject/nx-libs" <nx-libs@noreply.github.com>
Cc: "Ulrich Sibiller" <uli42@gmx.de>; "Mention" <mention@noreply.github.com>
Betreff: Re: [ArcticaProject/nx-libs] autograb (without inputlock, for now)(#494)

On Mo 24 Jul 2017 15:12:06 CEST, Ulrich Sibiller wrote:

> Activate autograb, Switch to Fullscreen using keystroke, Switch back 
> . Try the same with switch-all-screens.
>
> For one of those autograb ist broken after wards.

@uli42: No, it is not. Tried that using nested i3 desktops. Works as 
expected here.

Just tried it once more. Can't trigger any sort of failure.

Mike
-- 

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22 0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

**Comment by sunweaver** [comment]
On  Mo 24 Jul 2017 15:25:00 CEST, Ulrich Sibiller wrote:

> So its kwin again... That is even worse. It should work with all WMs equally.
>
>> For one of those autograb ist broken after wards.

Sooo... what do you mean by "broken afterwards"? The toggling doesn't  
work anymore? Or autograb is disabled whereas it was enabled before  
switching fullscree/allscreens?

Details welcome! You can I reproduce the issue?

Mike
-- 

DAS-NETZWERKTEAM
mike gabriel, herweg 7, 24357 fleckeby
mobile: +49 (1520) 1976 148
landline: +49 (4354) 8390 139

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

**Comment by uli42** [comment]
Ok, more details:
1. use kwin or openbox on Ubuntu 16.04 lTS. Run nxagent :66. Activate autograb, it is working fine.
2a. press fullscreen keystroke -> does not work for, makes window disappear. Press fullscreen keystroke again -> window is back
2b. press allscreens keystroke -> works. Press allscreens keystroke again -> windowed mode is back  again. But the window is not active anymore, autograb is not working. You have to activate the window (again) to make autograb work

If you add the patch from #458 2a) works better. However,while in fullscreen you can still press alt-tab to change windows and afterwards the fullscreen keystroke is not working anymore. You have to use alt-tab to (somehow) activate the shell you started nxagent from and then (invisibly, because the fullscreen nxagent is on top and you see only a black screen) press ctrl-c. 

So please try to test this with kwin. I think you will agree that we cannot ship it like that...

**Comment by sunweaver** [comment]
I can reproduce the issue with openbox runing on the hosting (local) X.org. Will investigate further, later.

File: nx-X11/programs/Xserver/hw/nxagent/Dialog.h [hunk 1]
```diff
@@ -212,5 +218,29 @@ Ctrl+Alt+E to enable it again.\
 
 #define DIALOG_DISABLE_DEFER_MODE_LOCAL 0
 
+
+#define DIALOG_ENABLE_AUTOGRAB_MODE_MESSAGE \
+\
+"\
+Keyboard auto-grabbing mode is now enabled. You can press\n\
```
Comment (uli42) [inline nx-X11/programs/Xserver/hw/nxagent/Dialog.h:227]: I don't like that hardcoded keystroke here. Maybe change the text to "press the same keystroke again to..."

**Comment by sunweaver** [comment]
@uli42: I have rebased the autograb-without-inputlock branch, can you give this another try? I can't remember the smoke test to get nxagent burnt (some switching between ctrl-F, ctrl-shift-F and windowed mode. Do you remember? (And with KWin running?)

**Comment by uli42** [comment]
It is described within this very issue

(fun fact: I have rebased my autograb branch just today, too)

**Comment by sunweaver** [comment]
On  Do 27 Jun 2019 22:47:03 CEST, Ulrich Sibiller wrote:

> It is described within this very issue
>
> (fun fact: I have rebased my autograb branch just today, too)

I won't be much at keyboard over the weekend and have Debian stuff  
todo tomorrow. Let me know if you get some to some testing over the  
next few days...

Thanks,
Mike
-- 

DAS-NETZWERKTEAM
c\o Technik- und Ökologiezentrum Eckernförde
Mike Gabriel, Marienthaler str. 17, 24340 Eckernförde
mobile: +49 (1520) 1976 148
landline: +49 (4351) 486 14 27

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

**Comment by uli42** [comment]
I am not sure about my weekend schedules right now. Will decide tomorrow.

However, just tested:
1. run NX_CLIENT=<path_to_nxdialog> nxagent :55
2. xterm -display :55
3. in xterm: startlxde
4. activate autograb -> works
5. ctrl-alt-f -> fullscreen
6. press alt-tab -> fullscreen will vanish while the alt-tab-switcher of your local X gets activated

**Comment by uli42** [comment]
https://gist.github.com/uli42/e246c4bfefeb62bccb961f768c4809d4 makes it almost work. When leavin fullscreen mode the window sometimes iconifies.

**Comment by sunweaver** [comment]
Hi,

On  Do 27 Jun 2019 23:51:20 CEST, Ulrich Sibiller wrote:

> https://gist.github.com/uli42/e246c4bfefeb62bccb961f768c4809d4 makes  
> it almost work. When leavin fullscreen mode the window sometimes  
> iconifies.

please feel free to add commits to the PR branch in my (sunweaver)  
github namespace.

Thanks+Greets,
Mike
-- 

DAS-NETZWERKTEAM
c\o Technik- und Ökologiezentrum Eckernförde
Mike Gabriel, Marienthaler str. 17, 24340 Eckernförde
mobile: +49 (1520) 1976 148
landline: +49 (4351) 486 14 27

GnuPG Fingerprint: 9BFB AEE8 6C0A A5FF BF22  0782 9AF4 6B30 2577 1B31
mail: mike.gabriel@das-netzwerkteam.de, http://das-netzwerkteam.de

**Comment by sunweaver** [comment]
Superceded by #835.

**Pull request closed without merging**